### ISSUE TITLE
[Nvidia] Black screen when HDR option is enabled

### ISSUE THREAD

**Issue by Algorithm0** [Plagman/gamescope#1593]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Are you using any gamescope patches or a forked version of gamescope?

- [X] The issue occurs on upstream gamescope without any modifications

### Current Behavior

When I activate the HDR option, I see that nothing is displayed on my TV in terms of images (there is sound, I hear Steam starting up and I return to the desktop by the sound of the keys).
To run Steam on my TV I use [ChimeraOS/gamescope-session](https://github.com/ChimeraOS/gamescope-session).
I will describe what I see in some steps of reproducing the problem:
3. I see Steam on my TV. I see the TV showing that it is currently outputting HDR. I see a bar flashing at the bottom of my screen.
4. The entire image is terribly distorted, image fragments are mixed up, the screen flickers.
6. I don't see anything. The TV shows a black screen. There is no HDR on/off sign either. There is only sound.

### Steps To Reproduce

1. Specify in the `gamescope-session-plus.conf` parametr `ENABLE_GAMESCOPE_HDR=1`;
2. Log out of the desktop session;
3. Login to "Steam Big Picture" session;
4. In Steam, go to display settings and enable HDR;
5. Select "Exit" and "Go to Desktop" (maybe there are other names on steam, I use another language, sorry);
6. Login to "Steam Big Picture" session;

### Hardware information

```markdown
- Distro: Fedora Linux 40 (Workstation Edition)
- CPU: AMD Ryzen™ 7 7800X3D
- GPU: NVIDIA GeForce RTX™ 4070 Ti
- Driver Version: 560.35.03
```

### Software information

```markdown
- Desktop environment: GNOME 46
- Session type: wayland
- Gamescope version: gamescope version 3.15.13 (gcc 14.2.1)
- Gamescope launch command(s): /usr/local/bin/gamescope -W 3840 -H 2160 --hdr-enabled --hdr-itm-enable --prefer-output HDMI-A-2 --xwayland-count 2 --default-touch-mode 4 --hide-cursor-delay 3000 --fade-out-duration 200 --steam -R /run/user/1000/gamescope.Mn36AFM/startup.socket -T /run/user/1000/gamescope.Mn36AFM/stats.pipe --prefer-vk-device 10de:2782
```

### Which gamescope backends have the issue you are reporting?

- [ ] Wayland (default for nested gamescope)
- [X] DRM (default for embedded gamescope, i.e. gamescope-session)
- [ ] SDL
- [ ] OpenVR

### Logging, screenshots, or anything else

[Logs after entering a session after enabling HDR from the Steam settings](https://github.com/user-attachments/files/17509376/gamescope-stdout.log)

**Comment by Algorithm0** [comment]
I have uploaded some videos to [my drive](https://drive.google.com/drive/folders/1nN5EzuOHoB1k4hF7LUO-rxfntDlKXu-x?usp=sharing) with the appropriate titles so you can see what is going on.

**Comment by matte-schwartz** [comment]
Your logs show it failing to modeset for the resolution it's picking, which is probably when you get a black screen. The other part, that flickering, looks exactly like an issue I reported a few months ago https://forums.developer.nvidia.com/t/display-modes-above-2560x1440p-120hz-with-hdr-enabled-cause-flickering-corruption-within-gamescope-session/295314 but I did not create a dedicated issue here at the time. 

Seems like maybe there's some NV issue with HDR over a certain resolution and refresh rate, but I don't think it ever got logged for an internal NV bug report.

This is my 4090 with gamescope's DRM backend (gamescope-session) at 7680x2160p@120hz with the same issue you recorded
![IMG_2249](https://github.com/user-attachments/assets/93195381-1872-4711-a4bc-9714e92dde05)

**Comment by kisak-valve** [comment]
The first thing to check would be if the NVIDIA 565 release series behaves better.

**Comment by matte-schwartz** [comment]
already did, no change sadly https://forums.developer.nvidia.com/t/565-release-feedback-discussion/310777/16

**Comment by matte-schwartz** [comment]
It's much rarer, but it seems like the Wayland backend also has the possibility to be affected by the same flickering, even when the WSI layer is disabled with `ENABLE_GAMESCOPE_WSI=0` and HDR is **not** used. 

![image](https://github.com/user-attachments/assets/eabc9e44-14e5-4056-9707-dade82e84be1)

I *do* have the argument to allow Nvidia colorspace in my KWin environment left over from Gamescope testing yesterday. The flickering went away when I alt-tabbed out of full screen Gamescope and started again once tabbing back in. 

My specific launch command was `gamescope -f -h 2160 -w 7680 -r 120 --adaptive-sync` in Lutris and only started after entering and exiting the game (Star Citizen) a few times without quitting the launcher. Screenshots did not capture the flickering which makes me suspect maybe something at scanout? Will include a drm_info log as well

**Comment by SimpleHeuristics** [comment]
this also happens when started in nested mode, artifacting disappears however when you hit the super/meta key to bring up the app menu in KDE Plasma 6.2.3

**Comment by trkorecky** [comment]
I'm seeing this issue as well, seems to repro much more reliably with `--hdr-itm-enable` as well.

Factorio launch command in Steam:
`VKD3D_DISABLE_EXTENSIONS=VK_KHR_present_wait DXVK_HDR=1 gamescope -W 3440 -H 1440 -r 60 -e -f --hdr-enabled --hdr-itm-enable --adaptive-sync --force-grab-cursor -- %command%`

Escape from Tarkov Gamescope custom settings in Lutris:
`-r 144 --hdr-enabled --hdr-itm-enable --hdr-itm-target-nits 1100 --hdr-itm-sdr-nits 200 --adaptive-sync --force-grab-cursor --mangoapp`

Removing `--hdr-itm-enable` reliably fixes the corruption. Alternately rebooting seems to help and lets me play without corruption, though I typically play and then shut down afterwards so I haven't tried multiple game sessions.

Haven't had HDR issues in games that natively support it like Baldur's Gate 3, Forza Horizon 4, or Horizon Zero Dawn Remastered using:
`VKD3D_DISABLE_EXTENSIONS=VK_KHR_present_wait DXVK_HDR=1 gamescope -W 3440 -H 1440 -r 144 -e -f --hdr-enabled --adaptive-sync --force-grab-cursor -- %command%`

Running latest CachyOS with 6.12.1 kernel, KDE Plasma 6.2.4, NVIDIA 565.57.01 drivers on a 7950X3D and RTX 3080. Plugged into an Acer Predator X35 with DisplayPort.

**Comment by jeisom** [comment]
I'm seeing this as well. If hdr is enabled with  --hdr-enabled with or without " --hdr-itm-enable --hdr-debug-force-support" I get either a blackscreen or sometimes no screen with a game that has hdr.  FarCry5, CP2077, Horizon Zero Dawn Remaster and Ghost of Tsushima.  If I disable the hdr options and envs, I can run them.

If I set ENABLE_HDR_WSI=0 I was able to launch HZD:R.  I still can't play with HDR enabled in my little test as it is corrupt and crashes at the hdr calibrate menu though.

Nvidia 4080 (565.*)
Plasma 6.2.4


Full command I am trying to use:
"ENABLE_HDR_WSI=1 DXVK_HDR=1 DXVK_ENABLE_NVAPI=1 /usr/bin/gamescope --hdr-enabled --hdr-itm-enable --nested-refresh 75 --hide-cursor-delay 3000 --fade-out-duration 200 --hdr-debug-force-support -f -W 5120 -H 2160 -h 2160 -w 5120  -- %command%"

**Comment by matte-schwartz** [comment]
Can we re-title this thread to something like:
"[NVIDIA] - Black/flickering screen when HDR is enabled" to make this easier to track?

it's a bit more accurate of a title, given "The entire image is terribly distorted, image fragments are mixed up, the screen flickers." from the original report and follow up comments. A NVIDIA label would also be warranted, as this only happens under nvidia-drm

**Comment by SmithAF** [comment]
Also experiencing this issue. Happened when I tried to use HDR, although weirdly enough it's now happening with SDR now too.

```bash
#!/bin/bash
REFRESH_RATE=120
WIDTH=3840
HEIGHT=2160
exec gamescope --backend drm -W "$WIDTH" -w "$WIDTH" -H "$HEIGHT" -h "$HEIGHT" -r "$REFRESH_RATE" --framerate-limit "$REFRESH_RATE"  -e -- steam -tenfoot -steamos3
```
- Nvidia 3080 with 565.77
- Fedora 41

**Comment by matte-schwartz** [comment]
@SmithAF does it still happen if you fully reboot your PC and try again with SDR? I've found that once it starts going wrong (flickering, black screen, etc.) your **only** option is to reboot and then try launching Gamescope again after the fresh reboot, even if only using SDR.

**Comment by SmithAF** [comment]
@matte-schwartz that seems to mostly work, although comes back again after starting starting up the session again without reboot.  With SDR it's only roughly 5% of the screen at the bottom that gets corrupted, but with HDR it's like the images above. So likely this SDR issue I'm seeing is only tangentially related.

**Comment by matte-schwartz** [comment]
@SmithAF thanks for the confirmation on that. from my own testing, if you leave HDR enabled or leave HDR disabled, then it should mostly work correctly **as long as you remain in a single gamescope-session**. if you leave the session and re-enter, **or** if you touch the HDR toggle or resolution drop-down menu at all in the DRM backend, it sends you into the fun flickering fiesta until you fully reboot your PC. 

I am no closer to figuring out _why_ this happens however, or why it very rarely spreads to nested gamescope on Plasma 6 Wayland. `drm_info` doesn't change at all between these good and bad states, so I can only assume something is happening internally within nvidia-drm that causes this.

as an aside, @kisak-valve friendly ping on: https://github.com/ValveSoftware/gamescope/issues/1593#issuecomment-2550018899, NVIDIA bug label was something I discussed with @Plagman a few weeks ago. feel free to reach out if you want to discuss it further, thanks!

**Comment by jarrard** [comment]
I've been getting this issue lately. Quite frustrating.   With 3.16.11/2

LMAO, I found a HACK around fix.  Basically make a plasma window rule to **remove titlebar, and force fullscreen off..**

Basically if you set the game window to not fullscreen the issue went away for me, or unfocused mode.  

Others should give this a try.  If I set window to FULLSCREEN again via plasma, it glitches out.

**Comment by SimpleHeuristics** [comment]
This visual corruption in Gamescope session appears to be resolved on the latest Bazzite builds that are including the 575.57.08 new feature branch drivers and with the current steam client as of writing this post.

I've gone through multiple reboots now and switching between desktop and gamescope session without this issue popping up at a resolution of 4K, HDR, and VRR enabled. 

I'm not sure if this was resolved with the driver or with a recent update to the steam client.

**Comment by jarrard** [comment]
Running that driver version here, still happens. At least there is a workaround for time being.

**Comment by bernhardberger** [comment]
Still an issue with the 590 drivers. I can basically confirm all of the findings reported in this thread.

Cold boot -> fine until a res change or other display related changes (HDR e.g.). Switching to Desktop and then back to gamescope-session -> broken. Startung gamescope in a new tty -> broken.

The super weird thing that I don't understand is: why is it working fine on a cold boot?

Somehow this feels like it may be an initialization or timing problem of some sort..?

Only happens at 4k120 resolution for me and only on nvidia.

**Comment by adamdmoss** [comment]
> The super weird thing that I don't understand is: why is it working fine on a cold boot?

It's odd.  But FYI the semi-workaround is to unload and reload the nvidia kernel modules after every gamescope session.  (And yes unfortunately that implies that you have to stop gdm+desktop too, i.e. `init 4` , making it only barely more convenient than a whole reboot.)

**Comment by matte-schwartz** [comment]
Last I heard NVIDIA was investigating this issue after reproducing it fwiw

**Comment by asheriif** [comment]
Anyone had any luck with this? It's still broken for me with the latest drivers

**Comment by jarrard** [comment]
Think driver 580 and onward fixed it for me. 

Also you may still need the NVIDIA HDR patch vk-hdr-layer-kwin6-git and variable for fully to work under NVIDIA.  

Not sure how long that will be required for however. 

Problem is I run close to bleeding edge with CachyOS while many other people are running older stable release distros which can be as much as 12months off from any fixes to this issue.

**Comment by asheriif** [comment]
@jarrard at 4K? I'm on latest everything too (testing on CachyOS as well) and 4K@144hz still results in fullscreen artifacts and glitches and is completely broken. 1440p is fine. Also the layer patch is no longer necessary with the latest drivers I believe.

**Comment by Xeeynamo** [comment]
> Anyone had any luck with this? It's still broken for me with the latest drivers

595.58.03, still broken. The only solution I found was to re-load the nvidia drivers before starting gamescope on DRM. I created myself the following script I launch on a separate TTY, and never had issues since:

```shell
sudo systemctl stop plasmalogin.service # needed to unload drivers
sleep 2
sudo modprobe -rf nvidia_uvm nvidia_drm nvidia_modeset nvidia
sudo modprobe nvidia
sudo modprobe nvidia_modeset
sudo modprobe nvidia_drm
sudo modprobe nvidia_uvm
sudo nvidia-smi -pm 1
export VKD3D_DISABLE_EXTENSIONS=VK_KHR_present_id,VK_KHR_present_wait # I don't remember if this is required
export STEAM_GAMESCOPE_VRR_SUPPORTED=1 # !!! only if you have VRR
sleep 3 && gamescopectl adaptive_sync_ignore_overlay 1 &
gamescope --steam --rt --immediate-flips \
  --output-width 3840 --output-height 2160 \
  --hide-cursor-delay \ # if you want to navigate with gamepad only
  --nested-refresh 144 \ # customize this to your refresh rate
  --hdr-enabled --hdr-debug-force-output \ # if you have HDR
  --prefer-vk-device "10DE:2C02" \ # only if you have multiple GPUs
  --prefer-output=HDMI-A-1 \ # only if you have multiple monitors
  -- steam -gamepadui -steamos3
```

**Comment by asheriif** [comment]
Thanks for sharing! Will try that once I can. Does this log you out of the plasma session on your main tty?

**Comment by Xeeynamo** [comment]
> Thanks for sharing! Will try that once I can. Does this log you out of the plasma session on your main tty?

That is correct. Terminating Plasma seems to be a hard requirement for `modprobe -rf` to work on the nvidia drivers (I updated the script above, I forgot to add that line).

**Comment by SimpleHeuristics** [comment]
> Last I heard NVIDIA was investigating this issue after reproducing it fwiw 

@matte-schwartz any idea if there has been any progress with this? On Bazzite 43 with their Gamescope session things seemed to be working very well could switch between desktop and Gamescope without issue even with 4K HDR and VRR. That version of Bazzite seemed to have a patched version of Gamescope. 

The new Bazzite 44 builds are closer to the upstream Gamescope and with that this issue seems to still be there.

**Comment by matte-schwartz** [comment]
I haven't heard anything about this driver bug since late last year, nor am I aware of what changes in a patched gamescope would affect the frequency of the issue.

**Comment by ManEnthusiast** [comment]
> I haven't heard anything about this driver bug since late last year, nor am I aware of what changes in a patched gamescope would affect the frequency of the issue.

I can verify the same experience.

The problem did mostly go away for several months on Bazzite but has only returned as of Bazzite 44.

Problem is that 44 changes so many things that it would be difficult to identify which specific change has caused it.

**Comment by SwiftYeeter10** [comment]
I didn't think when I googled this it would turn out to be an unresolved 2 year old issue, damn.

**Comment by matte-schwartz** [comment]
I've updated my report to NVIDIA with my results from debugging the issue today, and there is absolutely no way gamescope can mitigate this. The fix will have to be on the NVIDIA driver side: https://forums.developer.nvidia.com/t/display-modes-above-2560x1440p-120hz-with-hdr-enabled-cause-flickering-corruption-within-gamescope-session/295314/27

TL;DR: the NVIDIA driver does weird stuff when scanning out images that are non-contiguous, and as the amount of non-contiguous memory gets filled by subsequent modesets and HDR toggles, the corruption compounds and the driver scans out random other stuff in its non-contiguous memory (like extra overlay copies). For now, you can build the NVIDIA open driver with https://github.com/NVIDIA/open-gpu-kernel-modules/pull/1305 and set something like:
```
options nvidia NVreg_RegistryDwords=RMDisableNoncontigAlloc=1
```
in `/etc/modprobe.d/nvidia-contig.conf`

but this is strictly a workaround.

**Comment by SimpleHeuristics** [comment]
> Replying to https://github.com/ValveSoftware/gamescope/issues/1593#issuecomment-5336562334

Would your patch not be the intended behaviour if the driver acts in a contiguous manner for memory allocation in the desktop compositors?

Either way amazing that you found this I hope it gets rolled into the next update driver but unfortunately seems a bit too niche for them to prioritize.

I tried to get this working in Bazzite last night but wasn't able to but that's likely more of an issue with my lack of experience.

**Comment by NightHammer1000** [comment]
https://github.com/NightHammer1000/gamescope/tree/poc/gamescope-gbm-route

This branch, together with forced composition sidesteps the Issue completly without the need to patch the Driver.

After days of testing, I found no Issues.
Matts Patch to the Driver caused multiple crashes in Vulkan Games on my end, while the GBM Gamescope Route did not.
A/B Testing also found no Performance Degradation.

Again, not a fix. Still a workaround from an diffrent angle.

**Comment by Mojobones** [comment]
> https://github.com/NightHammer1000/gamescope/tree/poc/gamescope-gbm-route
> 
> This branch, together with forced composition sidesteps the Issue completly without the need to patch the Driver.
> 
> After days of testing, I found no Issues. Matts Patch to the Driver caused multiple crashes in Vulkan Games on my end, while the GBM Gamescope Route did not. A/B Testing also found no Performance Degradation.
> 
> Again, not a fix. Still a workaround from an diffrent angle.

I just gave this branch a shot, and it worked! great workaround until the drivers themselves are fixed (whenever that may be)

**Comment by cubanismo** [comment]
FWIW, the language here is sort of backwards. There are various issues:

* The hardware can not display non-contiguous buffers. That isn't considered a bug, it's just the way the hardware is.
* The driver doesn't enforce that limitation at atomic commit nor DRM FB creation time. That's a driver bug, and we'll fix that, but it would just make the failure more obvious for Gamescope users. The desired behavior still wouldn't be achieved.
* Gamescope allocates its scan-out surfaces with Vulkan, which has no way to specify the memory is meant for display use. That's a bug in Gamescope, but the Gamescope developers have indicated they are not going to fix that, or at least are not going to fix it by using a display-capable allocation API like GBM. This is not without reason, but I don't want to try to convey their reasoning and get it wrong, so I'll let them clarify if they choose to.

So allocating with GBM is actually the most correct fix proposed yet, but it seems it is not acceptable. Given that, we'll try to implement a work-around in the driver, and probably make it specific to Gamescope, since all the heuristics for the work-around I've heard so far have downsides that could impact performance if applied generally. Separately, I'd like to enhance Vulkan to support allocating displayable surfaces properly, but that's a surprisingly large extension when done correctly, so it likely will not happen soon.

**Comment by NightHammer1000** [comment]
Thank you. Thats really great to hear.

**Comment by moi952** [comment]
> Replying to https://github.com/ValveSoftware/gamescope/issues/1593#issuecomment-5402337747

That would be really great! 
Thanks for the info—we’ve been waiting for this for so long. 
Nvidia drivers are improving, but Game Mode is very tricky due to those artifacts. 

Will you also take this opportunity to fix the transparency issues that occur when opening Steam overlays and notifications? 

And do you have an ETA? 
Thanks

**Comment by NightHammer1000** [comment]
Info to the Transparency Issue here.
https://github.com/ValveSoftware/steam-for-linux/issues/13372#issuecomment-5372228331

Might be out of scope for a Driver tho

**Comment by cubanismo** [comment]
Yes, to avoid muddying this topic, I won't comment on unrelated issues here. If you want my attention (Or someone else at NVIDIA's attention) on a given issue, please @ me there, rather than pinging me wherever I show up. That said, I don't have time nor expertise to respond to everything, so if I don't respond when mentioned, I'm probably either overloaded at the moment, or don't have anything to add on the topic.

**Comment by Algorithm0** [comment]
<details><summary>

What really seems funny to me is that I personally don't experience the configuration at the moment. At the moment, I use my own fork from the [gamescope+](https://github.com/Algorithm0/gamescope-session) project. I am using the following config:

</summary>

```sh
# Size of the screen. If not set gamescope will detect native resolution from drm.
#SCREEN_HEIGHT=2160
#SCREEN_WIDTH=3840
#export STEAM_DISPLAY_REFRESH_LIMITS=60,120

GAMESCOPE_BIN="/usr/local/bin/gamescope"

#GAMESCOPE_BIN="/home/steam-config/gamescope/build/3.15.12/src/gamescope"
#GAMESCOPE_BIN="/home/steam-config/gamescope/build/3.15.13/src/gamescope"
#GAMESCOPE_BIN="/home/steam-config/gamescope/build/3.15.14/src/gamescope"
#GAMESCOPE_BIN="/home/steam-config/gamescope/build/3.16.2/src/gamescope"

BEFORE_GAMESCOPE_SESSION_PLUS="/home/steam/.config/environment.d/tv_on.py;"
AFTER_GAMESCOPE_SESSION_PLUS="/home/steam/.config/environment.d/tv_off.py"
#BEFORE_GAMESCOPE_CLIENT="/bin/bash /home/steam/.config/environment.d/alice_connect.sh"

#DISABLE_GAMESCOPE_WSI=1

BACKEND=drm
#BACKEND=wayland
#DRM_MODE=fixed

ENABLE_GAMESCOPE_HDR=1
GAMESCOPE_HDR_NITS=700

#OUTPUT_CONNECTOR=HDMI-A-2,DP-4
#OUTPUT_CONNECTOR=DP-4
OUTPUT_CONNECTOR=HDMI-A-2
```

</details>

So I run steam on my TV as a headless session. I get both HDR and 120 fps, and 4k.

<details><summary>I'll leave my software version here as well:</summary>

```sh
aurum@fedora:~$ gamescope --version
[gamescope] [Info]  console: gamescope version 3.16.25 (clang 22.1.8)
aurum@fedora:~$ nvidia-smi
Wed Aug 26 23:29:25 2026
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 610.57.04              KMD Version: 610.57.04     CUDA UMD Version: 13.3     |
+-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA GeForce RTX 4070 Ti     Off |   00000000:01:00.0  On |                  N/A |
| 49%   74C    P0            236W /  285W |    5890MiB /  12282MiB |     98%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+

+-----------------------------------------------------------------------------------------+
| Processes:                                                                              |
|  GPU   GI   CI              PID   Type   Process name                        GPU Memory |
|        ID   ID                                                               Usage      |
|=========================================================================================|
|    0   N/A  N/A            2417    C+G   /usr/local/bin/gamescope                152MiB |
|    0   N/A  N/A            2454      G   Xwayland                                  3MiB |
|    0   N/A  N/A            2455      G   Xwayland                                  3MiB |
|    0   N/A  N/A            2571      G   ...share/Steam/ubuntu12_32/steam          7MiB |
|    0   N/A  N/A            2886      G   ./steamwebhelper                         21MiB |
|    0   N/A  N/A            2909      G   ...am/ubuntu12_64/steamwebhelper        295MiB |
|    0   N/A  N/A           45762      G   C:\windows\system32\explorer.exe          3MiB |
|    0   N/A  N/A           45876    C+G   ...\common\PRAGMATA\PRAGMATA.exe       4953MiB |
+-----------------------------------------------------------------------------------------+
aurum@fedora:~$ cat /etc/fedora-release
Fedora release 44 (Forty Four)
```

<img width="4080" height="3072" alt="Image" src="https://github.com/user-attachments/assets/d0575ba4-a9d9-4a35-92a1-271c7132186f" />
<img width="4080" height="3072" alt="Image" src="https://github.com/user-attachments/assets/7e13c255-0b0f-4ba6-994d-bf5832bf9d13" />

</details>
